NiranjanBhat123/major_project
# Integrate deepface face recognition model, API and its endpoints

# Summary
This PR introduces the integration of the DeepFace face recognition model within our application, adding a new API endpoint (/verify_faces) that allows for face verification between two images. This feature leverages the DeepFace library, specifically the VGG-Face model, to verify if two images contain matching faces, returning similarity scores and match confidence.
 

# Detailed Description 
The face verification process compares two facial images and determines whether they match based on the model’s similarity calculations. The integration is implemented with two main methods:

- Base64 Image Verification - Supports base64-encoded images by decoding them and passing them directly to the DeepFace verification function.
- InMemoryUploadedFile Image Verification - Supports standard file uploads in the form of InMemoryUploadedFile objects, allowing users to upload images directly for comparison.
How the DeepFace Model Works

- DeepFace is a facial recognition framework that leverages pre-trained deep learning models for facial feature extraction and matching. In this implementation:

- Model: The VGG-Face model, a convolutional neural network trained on millions of facial images, is used. VGG-Face has been proven effective in recognizing faces in diverse scenarios with high accuracy.
- Face Detection: Before verifying the faces, the model checks if both images contain detectable faces using OpenCV’s Haar Cascade Classifier. This ensures that we only proceed with facial images, as non-facial images could lead to misinterpretation.
- Feature Matching: The core verification process involves comparing the feature vectors extracted from both images.
- DeepFace calculates the Euclidean or cosine distance between these vectors. Lower distances indicate higher similarity:
- Threshold: The model returns a threshold value specific to the VGG-Face model. If the calculated distance is below this threshold, the images are deemed a match; otherwise, they are not.
- Confidence Score: DeepFace also provides a verified flag and distance metrics. The confidence of the match is based on whether the distance metric falls within the threshold.
Workflow

## Request Structure:
- The /verify_faces endpoint accepts a POST request with two images: either as base64-encoded strings or file uploads.
### Verification Steps:
- The request data is processed to check for images in either base64 or file format.
- Each image undergoes face detection to ensure that both images contain faces.
- The images are passed to the DeepFace verify function, where similarity checks are performed, and the model returns a match status along with additional data such as the distance and threshold.
Response:
- If verification is successful, the endpoint returns a JSON response with success, matched, distance, and threshold.
- If any issues arise (e.g., no face detected or error in processing), an error message is returned.
Example Response

```
{
  "success": true,
  "matched": true,
  "distance": 0.45,
  "threshold": 0.50,
  "model": "VGG-Face"
}
```

## Testing 
- Manual Testing: Tested with sample images to validate the matching and non-matching cases.
- Postman: Verified CSRF exemption and checked responses for both base64 and file upload cases.
## Dependencies
- DeepFace: Added DeepFace library for face recognition.
- OpenCV: Used for face detection with Haar Cascades.
- NumPy: Required for image processing and decoding base64 images.
